capitulo2.md

```diff
@@ -222,3 +222,94 @@ int main()
 
 Assim como qualquer outro tipo de variável, é possível adicionar `Point` como parâmetro para funções, como é o caso do exemplo acima, na linha 9. A função `calculate_distance` necessita de dois parâmetros, `p1` e `p2`, ambos do tipo `Point` para ser invocada. Note que, ao invocar a função na linha 19, as duas variáveis `p1` e `p2` são copiadas para o escopo da função `calculate_distance`. Ou seja, se alterarmos os valores de `p1` e `p2` dentro do corpo da função `calculate_distance`, os valores de `p1` e `p2` presentes no escopo da função `main` não serão alteradas.
 
+
+## 2.5 Referências
+
+Referências permitem criar "apelidos" para variáveis, isto é, uma referência para uma variável acessa o valor daquela variável, mas nomeada de outra forma. Considere o exemplo abaixo, onde é criada uma referência de nome "b" para uma variável de nome "a". Utiliza-se o simbolo "&" junto ao tipo da variável para declarar uma referência. Diferente de uma variável "comum", alterar o valor da referência altera o valor da variável referenciada. Portanto, no exemplo abaixo, o resultado vai ser a mensagem `a = 7, b = 7`.
+
+```
+#include <iostream>
+ 
+int main()
+{
+    int a = 1;
+    int& b = a;
+ 
+    b = 7;
+    std::cout << "a = " << a << ", b = " << b << std::endl;
+
+    return 0;
+}
+```
+
+Como uma referência é um "apelido" para outra variável, não é possível criar uma referência para um valor (Por exemplo, `int& b = 9;` não faz sentido, e ocasiona em um erro de compilação). O uso de referências não é restrito ao caso de variáveis locais. Por exemplo, é possível invocar funções passando referências como parâmetros. É importante lembrar que, normalmente, os parâmetros da mesma são copiados para o escopo da função. Para tipos primitivos, isso não é um problema, porém, para estruturas mais complexas é possível que se tenha uma degradação de performance devido à cópias desnecessárias.
+
+Considere o exemplo abaixo. Imagine que `Matrix` é uma estrutura de dados que representa uma matriz 100x100. A função `sum_elements` soma todos os elementos dessa matriz. O loop na função `main` foi criado sintéticamente para representar muitas chamadas da função `sum_elements`. Para cada invocação da função `sum_elements`, a estrutura de dados da matriz é copiada para o contexto da função, apenas para somar o valor de cada um de seus elementos.
+
+```
+/**
+ * Sum of elements in a given 100x100 Matrix.
+ * :param Matrix matrix:
+ * :returns double:
+ */
+double sum_elements(Matrix matrix)
+{
+    double sum = 0.;
+    for (int i = 0; i < 100; ++i) {
+        for (int j = 0; j < 100; ++j) {
+            sum += matrix[i][j];
+        }
+    }
+    return sum;
+}
+ 
+int main()
+{
+    // ...
+    double total_sum = 0.;
+    for (int i = 0; i < 1000000; ++i) {
+        total_sum += sum_elements(matrix);
+    }
+    std::cout << "Total Sum = " << total_sum << std::endl;
+ 
+    return 0;
+}
```

Comment:
Seria legal o exemplo utilizar std::chrono para comprar tempos de execução e ser compilavel, talvez com link pro godbolt ?

Reply:
Então, o meu exemplo original tem exatamente isso (Eu tirei essa parte do material da aula que eu dou de C++), mas fiquei com dúvida pq nesse ponto teóricamente a pessoa "não sabe" STL ainda. To pensando em uma forma "linear" de aprendizado e de leitura... Será que coloco o exemplo com chrono mesmo assim? Talvez dê pra colocar e explicar "ó, da pra fazer isso dessa forma. Mais pra frente vamos ver o que essas coisas significam, mas por enquanto só acredita que ta certo -- para mostrar os resultados".

O que acham?

Reply:
Faz sentido, e oq vc acha do https://quick-bench.com/ ?

Reply:
Então, tentei aqui. Mas tem jeito de manter a URL?
```
#include <vector>

using Matrix = std::vector<std::vector<int>>;

double sum_elements(Matrix matrix)
{
    double sum = 0.;
    for (int i = 0; i < 100; ++i) {
        for (int j = 0; j < 100; ++j) {
            sum += matrix[i][j];
        }
    }
    return sum;
}
static void WithCopy(benchmark::State& state) {
  auto matrix = Matrix(100, std::vector<int>(100));

  for (auto _ : state) {
    sum_elements(matrix);
  }
}
BENCHMARK(WithCopy);

double sum_elements_ref(Matrix& matrix)
{
    double sum = 0.;
    for (int i = 0; i < 100; ++i) {
        for (int j = 0; j < 100; ++j) {
            sum += matrix[i][j];
        }
    }
    return sum;
}
static void WithRef(benchmark::State& state) {
  auto matrix = Matrix(100, std::vector<int>(100));

  for (auto _ : state) {
    sum_elements_ref(matrix);
  }
}
BENCHMARK(WithRef);
```

Reply:
![image](https://user-images.githubusercontent.com/1008957/135179219-41bdb09c-ac1e-41e3-a315-c374f62ab548.png)

Reply:
Com `-O0`

![image](https://user-images.githubusercontent.com/1008957/135179292-46622888-36f7-4c68-ad4d-5ae4e79b95f9.png)